Subsets

rathmann opened the pull request:
<!-- Optimize Itertools.subsets() to use return rather than yield from -->

#### References to other Issues or PRs
<!-- Fixes #23056 -->
Fixes #23056

#### Brief description of what is fixed or changed
When a Python function delegates to a generator or other Iterator, common practice is for the caller to be a generator and to use yield from. However, in certain simple circumstances, it is OK for the caller not to be a generator itself, and to just return the result (some kind of Iterator) of the call. This can save some overhead.


#### Other comments
This PR applies this technique to `itertools.subsets()` and  `itertools.variations()`.  

The function `subsets()` has been part of SymPy for quite some time, and is used in various places around the code base.   In these cases I suspect that the cost of `subsets()` itself is a fairly small part of the overall execution path, so I don't want to claim that end-users of that functionality will notice any significant difference.

Question for reviewers - does this need a comment in the body of the function explaining the approach?  I didn't change the docstring, since behavior should be unchanged.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* utilities
    * itertools.subsets() and itertools.variations() should be a bit faster.  The return type of these functions now depends on the arguments, but is always an iterator.  In particular, isinstance(result, collections.abc.Iterator) should always be True.
<!-- END RELEASE NOTES -->

rathmann committed:
Optimization of subsets() to use return rather than yield from

By avoiding the use of yield in the body of iterables.subsets, Python
sees it as just a regular function rather than a generator.  Hence it
can call generators and return the resulting generator objects,
avoiding some overhead from a layer of yield from handling.

rathmann committed:
Merge branch 'master' of git://github.com/sympy/sympy into subsets

rathmann committed:
Merge branch 'master' of https://github.com/sympy/sympy into subsets

oscarbenjamin commented:
This looks good to me. Do you want to also change the other functions you mention?

rathmann commented:
Thank you Oscar, and also @smichr for the thumbs up.   Yes, I think I would like to add an analogous change to variations (see linked issue for a prototype).

rathmann committed:
Reduce "yield from" overhead in iterables.variations()

Same treatment as was recently applied to iterables.subsets().  All
paths through iterables.variations() still return a generator object,
but variations() is no longer itself a generator function.  This
avoids the overhead of one level of yield from.  This change can save
as much as 30% of the time spent in this function.  (Very sensitive to
Python version.)

rathmann commented:
Thank you Oscar and Chris.

I have added the `variations()` changes to this PR.

I needed to come up with expressions for an empty generator object,
and a generator object which yields a single empty tuple.  If you have
better idioms for these than what I used, please let me know.

Also, I have been considering including a comment at the top of
variations() along the lines of

```# Not a generator function, but returns a generator object.```

So far I haven't, because I fear it would be more confusing than
helpful.  But, please let me know if you think some explanation would
be a good idea.

smichr commented:
From the user's point of view the function gives back something which must be enumerated/unpacked, and I don't think it matters (other than for performance) what that object is as long as it produces the same result.

e.g.
in master
```python
>>> variations([1,2,3],2)
<generator object variations at 0x03EA1670>
>>> list(_)
[(1, 2), (1, 3), (2, 1), (2, 3), (3, 1), (3, 2)]
>>> variations([],1)
<generator object variations at 0x03EA19F0>
>>> list(_)
[]
```
in this PR
```
>>> variations([1,2,3],2)
<itertools.permutations object at 0x021D41E0>
>>> list(_)
[(1, 2), (1, 3), (2, 1), (2, 3), (3, 1), (3, 2)]
>>> variations([],2)
<generator object variations.<locals>.<genexpr> at 0x007B49F0>
>>> list(_)
[]
>>>
```

smichr commented:
Regarding the null case, see [discussion here](https://stackoverflow.com/questions/13243766/python-empty-generator-function) which suggests `return iter(())` as a concise, Pythonic idiom:
```python
>>> def a():
...   return (_ for _ in [])
...
>>> def b():
...   return iter(())
...
>>> a()
<generator object a.<locals>.<genexpr> at 0x00698D30>
>>> list(_)
[]
>>> b()
<tuple_iterator object at 0x021D1B90>
>>> list(_)
[]
```

rathmann commented:
Thank you @smichr.   Excellent points.

I will change the 0 and 1 length iterators to `iter(())` and `iter(((),))`.   

And yes, the return type of variations and subsets is no longer always a  generator.   I am changing the docstring for variations not to promise a specific type.  I would think that users of a duck-typed language are not going to care that much about the precise return type.

Will push to github once tests fininsh.

rathmann commented:
I did a quick check, and these functions now return results of several types, but it is always the case that `isinstance(result, collections.abc.Iterator)` is True.

```python
python test_return.py 
Tests for Variations
['abc', 2, False] <class 'itertools.permutations'> True
['ab', 2, False] <class 'itertools.permutations'> True
['ab', 4, False] <class 'tuple_iterator'> True
['abc', 0, True] <class 'tuple_iterator'> True
['abc', 2, True] <class 'itertools.product'> True
Tests for Subsets
['abc', None, False] <class 'itertools.chain'> True
['abc', None, True] <class 'itertools.chain'> True
['abc', 0, False] <class 'itertools.combinations'> True
['abc', 1, False] <class 'itertools.combinations'> True
['abc', 2, False] <class 'itertools.combinations'> True
['abc', 3, False] <class 'itertools.combinations'> True
['abc', 4, False] <class 'itertools.combinations'> True
['abc', 0, True] <class 'itertools.combinations_with_replacement'> True
['abc', 1, True] <class 'itertools.combinations_with_replacement'> True
['abc', 2, True] <class 'itertools.combinations_with_replacement'> True
['abc', 3, True] <class 'itertools.combinations_with_replacement'> True
['abc', 4, True] <class 'itertools.combinations_with_replacement'> True
```

rathmann committed:
Update variations() docstring - don't claim result type is generator

Now just say it is an iterator.  Also, use different idiom for
0-length and 1-length results.

smichr commented:
This looks good. It's in.

rathmann commented:
Thank you so much, Christopher!
